src/components/CytoscapeGraph/Layout/BoxLayout.ts

```diff
@@ -0,0 +1,269 @@
+ */
+class SyntheticEdgeGenerator {
+  private nextId = 0;
+  private generatedMap = {};
+
+  public getEdge(source: any, target: any) {
+    const sourceId = this.normalizeToParent(source).id();
+    const targetId = this.normalizeToParent(target).id();
+
+    if (sourceId === targetId) {
+      return false;
+    }
+
+    const key = `${sourceId}->${targetId}`;
+
+    if (this.generatedMap[key]) {
+      return false;
+    }
+
+    this.generatedMap[key] = true;
+
+    return {
+      group: 'edges',
+      data: {
+        id: 'synthetic-edge-' + this.nextId++,
+        source: sourceId,
+        target: targetId
+      }
+    };
+  }
+
+  // Returns the parent if any or the element itself.
+  private normalizeToParent(element: any) {
+    return element.isChild() ? element.parent() : element;
+  }
+}
+
+/**
+ * Main class for the BoxLayout, used to bridge with cytoscape to make it easier to integrate with current code
+ */
+export default class BoxLayout {
+  readonly options;
+  readonly cy;
+  readonly elements;
+  readonly syntheticEdgeGenerator;
+
+  constructor(options: any) {
+    this.options = { ...options };
+    this.cy = this.options.cy;
+    this.elements = this.options.eles;
+    this.syntheticEdgeGenerator = new SyntheticEdgeGenerator();
+  }
+
+  /**
+   * This code gets executed on the cy.layout(...).run() is our entrypoint of this algorithm.
+   */
+  run() {
+    const { appBoxLayout, clusterBoxLayout, defaultLayout, namespaceBoxLayout } = this.options;
+    let allBoxNodes = this.cy.collection();
+    let removedElements = this.cy.collection();
+    let syntheticEdges = this.cy.collection();
+
+    for (const boxByType of [BoxByType.APP, BoxByType.NAMESPACE, BoxByType.CLUSTER]) {
+      const boxNodes = this.getBoxNodes(boxByType);
+      allBoxNodes = allBoxNodes.add(boxNodes);
+
+      let boxLayoutOptions = defaultLayout;
+      switch (boxByType) {
+        case BoxByType.APP:
+          if (appBoxLayout) {
+            boxLayoutOptions = appBoxLayout;
+          }
+          break;
+        case BoxByType.CLUSTER:
+          if (clusterBoxLayout) {
+            boxLayoutOptions = clusterBoxLayout;
+          }
+          break;
+        case BoxByType.NAMESPACE:
+          if (namespaceBoxLayout) {
+            boxLayoutOptions = namespaceBoxLayout;
+          }
+          break;
+      }
+
+      // (1.a) Prepare box node by assigning a size and running the compound layout
+      boxNodes.each(boxNode => {
+        const boxedNodes = boxNode.children();
+        const boxedElements = boxedNodes.add(boxedNodes.edgesTo(boxedNodes));
+        const boxLayout = boxedElements.layout(boxLayoutOptions);
+
+        // discrete layouts (dagre) stop before run() returns, continuous layouts (cose,cola) don't,
+        // force discrete behavior because we need each box to layout before continuing.
+        this.runDiscrete(boxLayout);
```

Comment:
This is not really waiting here. Is that OK?

Reply:
Should be fixed.